azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```diff
@@ -0,0 +1,151 @@
+            TWIN_DP_PATCH: None,
+            # Other handler tasks
+            # TODO: add
+        }
+
+        # Inbox handlers
+        self._on_message_received = None
+        self._on_method_request_received = None
+        self._on_twin_desired_properties_patch_received = None
+
+        # Other handlers
+        # TODO: add
+
+    async def _run_inbox_handler(self, inbox, handler_name):
+        """Run infinite loop that waits for an inbox to receive an object from it, then calls
+        the handler with that object
+        """
+        # Run the handler in a threadpool, so that it cannot block other handlers (from a different task),
+        # or the main client thread. The number of worker threads forms an upper bound on how many instances
+        # of the same handler can be running simultaneously.
+        tp = ThreadPoolExecutor(max_workers=4)
+        while True:
+            retval = await inbox.get()
+            # NOTE: we MUST use getattr here using the handler name, as opposed to directly passing
+            # the handler in order for the handler to be able to be updated without cancelling
+            # the running task created for this coroutine
+            handler = getattr(self, handler_name)
+            if inspect.iscoroutinefunction(handler):
+                # Wrap the coroutine in a function so it can be run in ThreadPool
+                def coro_wrapper(coro, arg):
+                    asyncio_compat.run(coro(arg))
+
+                tp.submit(coro_wrapper, handler, retval)
+            else:
+                # Run function directly in ThreadPool
+                tp.submit(handler, retval)
+
+    async def _run_event_handler(self, handler_name):
+        # TODO: implement
+        logger.error("._run_event_handler() not yet implemented")
+
+    def _get_inbox_for_handler(self, handler_name):
+        """Retrieve the inbox relevant to the handler"""
+        if handler_name == METHOD:
+            return self._inbox_manager.get_method_request_inbox()
+        elif handler_name == TWIN_DP_PATCH:
+            return self._inbox_manager.get_twin_patch_inbox()
+        elif handler_name == MESSAGE:
+            return self._inbox_manager.get_unified_message_inbox()
+        else:
+            return None
+
+    def _add_handler_task(self, handler_name):
+        """Create, and store a task for running a handler
+        """
+        # First check if the handler task already exists
+        if self._handler_tasks[handler_name] is not None:
+            raise AsyncHandlerManagerException(
+                "Cannot create task for handler: {}. Task already exists".format(handler_name)
+            )
+
+        inbox = self._get_inbox_for_handler(handler_name)
+
+        if inbox:
+            coro = self._run_inbox_handler(inbox, handler_name)
+        else:
+            coro = self._run_event_handler(handler_name)
+        task = asyncio_compat.create_task(coro)
+        self._handler_tasks[handler_name] = task
```

Comment:
>self._handler_tasks[handler_name] = task [](start = 8, length = 40)

What happens if the task raises an exception?  From what I've seen so far, it dies and becomes a dead husk that occupies the self._handler_tasks dictionary, creating exceptions and havoc for all eternity.  The task object has an add_done_callback that might be useful.

Reply:
Good catch. I'm going to do something with this in the next PR, because a similar problem (although with a different solution) exists for sync as well + i'm unifying some of the architecture between the two

---
In reply to: [456684624](https://github.com/Azure/azure-iot-sdk-python/pull/620#discussion_r456684624) [](ancestors = 456684624)